src/ciphers/morse_code.rs

```diff
@@ -0,0 +1,206 @@
+        "J" => ".---",    "K" => "-.-",     "L" => ".-..",
+        "M" => "--",      "N" => "-.",      "O" => "---",
+        "P" => ".--.",    "Q" => "--.-",    "R" => ".-.",
+        "S" => "...",     "T" => "-",       "U" => "..-",
+        "V" => "...-",    "W" => ".--",     "X" => "-..-",
+        "Y" => "-.--",    "Z" => "--..",
+
+        "1" => ".----",   "2" => "..---",   "3" => "...--",
+        "4" => "....-",   "5" => ".....",   "6" => "-....",
+        "7" => "--...",   "8" => "---..",   "9" => "----.",
+        "0" => "-----",
+
+        "&" => ".-...",   "@" => ".--.-.",  ":" => "---...",
+        "," => "--..--",  "." => ".-.-.-",  "'" => ".----.",
+        "\"" => ".-..-.", "?" => "..--..",  "/" => "-..-.",
+        "=" => "-...-",   "+" => ".-.-.",   "-" => "-....-",
+        "(" => "-.--.",   ")" => "-.--.-",  " " => "/",
+        "!" => "-.-.--",
+    }
+}
+
+fn _morse_to_alphanumeric_dictionary() -> HashMap<&'static str, &'static str> {
+    map! {
+        ".-"   =>  "A",      "-..." => "B",    "-.-." => "C",
+        "-.."  =>  "D",      "."    => "E",       "..-." => "F",
+        "--."  =>  "G",      "...." => "H",    ".." => "I",
+        ".---" =>  "J",     "-.-" => "K",     ".-.." => "L",
+        "--"   =>  "M",       "-." => "N",      "---" => "O",
+        ".--." =>  "P",     "--.-" => "Q",    ".-." => "R",
+        "..."  =>  "S",      "-" => "T",       "..-" => "U",
+        "...-" =>  "V",     ".--" => "W",     "-..-" => "X",
+        "-.--" =>  "Y",     "--.." => "Z",
+
+        ".----" => "1",    "..---" => "2",   "...--" => "3",
+        "....-" => "4",    "....." => "5",   "-...." => "6",
+        "--..." => "7",    "---.." => "8",   "----." => "9",
+        "-----" => "0",
+
+        ".-..." => "&",    ".--.-." => "@",  "---..." => ":",
+        "--..--" => ",",   ".-.-.-" => ".",  ".----." => "'",
+        ".-..-." => "\"",  "..--.." => "?",  "-..-." => "/",
+        "-...-" => "=",   ".-.-." => "+",   "-....-" => "-",
+        "-.--." => "(",   "-.--.-" => ")",  "/" => " ",
+        "-.-.--" => "!",  " " => " ",       "" => ""
+    }
+}
+
+fn _check_part(string: &str) -> bool {
+    for c in string.chars() {
+        match c {
+            '.' | '-' | ' ' => (),
+            _ => return false,
+        }
+    }
+    true
+}
+
+fn _check_all_parts(string: &str) -> bool {
+    string.split('/').all(_check_part)
+}
+
+fn _decode_token(string: &str) -> String {
+    _morse_to_alphanumeric_dictionary()
+        .get(string)
+        .unwrap_or(&_UNKNOWN_MORSE_CHARACTER)
+        .to_string()
+}
+
+fn _decode_part(string: &str) -> String {
+    string
+        .split(' ')
+        .map(_decode_token)
+        .collect::<Vec<String>>()
+        .join("")
+}
+
+/// Convert morse code to ascii.
+///
+/// Given a morse code, return the corresponding message.
+/// If the code is invalid, the undecipherable part of the code is replaced by `_`.
+pub fn decode(string: &str) -> Result<String, io::Error> {
```

Comment:
Hmmm, it says that this is never used 👀

Reply:
Which is weird because it is used  in a test.

Reply:
Yeah, I think that it is because the compiler struggles to get the references when stuff is not explicitly imported anywhere. Would you mind trying that out? It's the comment below.

Reply:
well either we figure that one out or just leave it there since the compiler is probably just confused, but the decoder is being properly tested.

Reply:
Found a solution